lang/cs/Org.Apache.REEF.Wake/Avro/ProtocolSerializer.cs

```diff
@@ -0,0 +1,175 @@
+﻿// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Reflection;
+using Microsoft.Hadoop.Avro;
+using Org.Apache.REEF.Utilities.Logging;
+using org.apache.reef.wake.avro.message;
+
+namespace Org.Apache.REEF.Wake.Avro
+{
+    /// <summary>
+    /// Given a namespace of Avro messages which represent a protocol, the ProtocolSerailizer
+    /// class reflects all of the message classes and builds Avro seriailziers and deserializers.
+    /// Avro messages are then serialized/deserilaized to and from byte[] arrays using the
+    /// Read/Write methods. A transport such as a RemoteObserver using a ByteCodec can then
+    /// be used to send and receive the serialized messages.
+    /// </summary>
+    public sealed class ProtocolSerializer
+    {
+        private static readonly Logger Logr = Logger.GetLogger(typeof(ProtocolSerializer));
+
+        // Delagates for message serializers and deserializers.
+        private delegate void Serialize(MemoryStream stream, object message);
+        private delegate void Deserialize(MemoryStream stream, object observer, long sequence);
+
+        // Message type to serialize/derserialize delagate.
+        private SortedDictionary<string, Serialize> serializeMap = new SortedDictionary<string, Serialize>();
+        private SortedDictionary<string, Deserialize> deserializeMap = new SortedDictionary<string, Deserialize>();
+
+        /// <summary>
+        /// Register all of the protocol messages using reflection.
+        /// </summary>
+        /// <param name="assembly">The Assembley object which contains the namespace of the message classes.</param>
+        /// <param name="messageNamespace">A string which contains the namespace the protocol messages.</param>
+        public void Initialize(Assembly assembly, string messageNamespace)
+        {
+            MethodInfo registerInfo = typeof(ProtocolSerializer).GetMethod("Register", BindingFlags.Instance | BindingFlags.NonPublic);
```

Comment:
This has moved to the constructor. Assembly and namespace must be set by the user.